AlignmentRecordRDD does not extend GenomicRDD per javac

The following Java source does not compile

``` java
  SparkContext sc = ...;
  JavaADAMContext javaAdamContext = new JavaADAMContext(sc);
  AlignmentRecordRDD alignments = javaAdamContext.loadAlignments(inputPath);
  JavaRDD<AlignmentRecord> rdd = alignments.jrdd();
```

``` bash
$ mvn compile
...
[ERROR] .../JavaCountAlignments.java:87: error: cannot find symbol
[ERROR]         JavaRDD<AlignmentRecord> rdd = alignments.jrdd();
[ERROR]                                                  ^
[ERROR]   symbol:   method jrdd()
[ERROR]   location: variable alignments of type AlignmentRecordRDD
```

<br/>
It appears javac does not consider `AlignmentRecordRDD` extending from `GenomicRDD`

``` bash
$ javap -cp adam-core_2.10-0.19.1-SNAPSHOT.jar \
  org.bdgenomics.adam.rdd.read.AlignmentRecordRDD

Compiled from "AlignmentRecordRDD.scala"
public interface org.bdgenomics.adam.rdd.read.AlignmentRecordRDD {
  public abstract org.bdgenomics.adam.rdd.fragment.FragmentRDD toFragments();
  ...

$ javap -cp adam-core_2.10-0.19.1-SNAPSHOT.jar \
  org.bdgenomics.adam.rdd.GenomicRDD

Compiled from "GenomicRDD.scala"
public interface org.bdgenomics.adam.rdd.GenomicRDD<T, U extends org.bdgenomics.adam.rdd.GenomicRDD<T, U>> {
  public abstract org.apache.spark.rdd.RDD<T> rdd();
  ...
```

Here is the hierarchy in Scala source

``` scala
sealed trait AlignmentRecordRDD
  extends AvroReadGroupGenomicRDD[AlignmentRecord, AlignmentRecordRDD]
{ ... }

abstract class AvroReadGroupGenomicRDD[T <% IndexedRecord: Manifest, U <: AvroReadGroupGenomicRDD[T, U]]
  extends AvroGenomicRDD[T, U]
{ ... }

abstract class AvroGenomicRDD[T <% IndexedRecord: Manifest, U <: AvroGenomicRDD[T, U]]
  extends ADAMRDDFunctions[T] with GenomicRDD[T, U]
{ ... }
```

<br/>
Surprisingly, this compiles and works at runtime

``` java
  JavaADAMContext javaAdamContext = new JavaADAMContext(sc);
  AlignmentRecordRDD alignments = javaAdamContext.loadAlignments(inputPath);
  JavaRDD<AlignmentRecord> rdd = ((GenomicRDD) alignments).jrdd();
```

``` bash
$ ADAM_MAIN=com.github.heuermh.adam.commands.JavaCountAlignments \
  adam-submit --jars adam-commands_2.10-0.19.1-SNAPSHOT.jar \
  -- \
  small.sam 

Using ADAM_MAIN=com.github.heuermh.adam.commands.JavaCountAlignments
Using SPARK_SUBMIT=/usr/local/bin/spark-submit
(1,20)
```

--- a/adam-cli/src/main/scala/org/bdgenomics/adam/cli/Transform.scala
+++ b/adam-cli/src/main/scala/org/bdgenomics/adam/cli/Transform.scala
@@ -27,7 +27,7 @@ import org.bdgenomics.adam.models.SnpTable
 import org.bdgenomics.adam.projections.{ AlignmentRecordField, Filter }
 import org.bdgenomics.adam.rdd.ADAMContext._
 import org.bdgenomics.adam.rdd.ADAMSaveAnyArgs
-import org.bdgenomics.adam.rdd.read.{ AlignedReadRDD, AlignmentRecordRDD }
+import org.bdgenomics.adam.rdd.read.AlignmentRecordRDD
 import org.bdgenomics.adam.rich.RichVariant
 import org.bdgenomics.utils.cli._
 import org.bdgenomics.utils.misc.Logging
@@ -453,7 +453,7 @@ class Transform(protected val args: TransformArgs) extends BDGSparkCommand[Trans
     })
 
     // make a new aligned read rdd, that merges the two RDDs together
-    val newRdd = AlignedReadRDD(mergedRdd, mergedSd, mergedRgd)
+    val newRdd = AlignmentRecordRDD(mergedRdd, mergedSd, mergedRgd)
 
     // run our transformation
     val outputRdd = this.apply(newRdd)

--- a/adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala
+++ b/adam-core/src/main/scala/org/bdgenomics/adam/rdd/ADAMContext.scala
@@ -52,7 +52,7 @@ import org.bdgenomics.adam.projections.{
 import org.bdgenomics.adam.rdd.contig.NucleotideContigFragmentRDD
 import org.bdgenomics.adam.rdd.feature._
 import org.bdgenomics.adam.rdd.fragment.FragmentRDD
-import org.bdgenomics.adam.rdd.read.{ AlignedReadRDD, AlignmentRecordRDD, UnalignedReadRDD }
+import org.bdgenomics.adam.rdd.read.AlignmentRecordRDD
 import org.bdgenomics.adam.rdd.variant._
 import org.bdgenomics.adam.rich.RichAlignmentRecord
 import org.bdgenomics.adam.util.{ ReferenceContigMap, ReferenceFile, TwoBitFile }
@@ -110,11 +110,7 @@ object ADAMContext {
 
   implicit def fragmentsToReadsConversionFn(fRdd: FragmentRDD,
                                             rdd: RDD[AlignmentRecord]): AlignmentRecordRDD = {
-    if (fRdd.sequences.isEmpty) {
-      UnalignedReadRDD(rdd, fRdd.recordGroups)
-    } else {
-      AlignedReadRDD(rdd, fRdd.sequences, fRdd.recordGroups)
-    }
+    AlignmentRecordRDD(rdd, fRdd.sequences, fRdd.recordGroups)
   }
 
   // Add ADAM Spark context methods
@@ -604,7 +600,7 @@ class ADAMContext(@transient val sc: SparkContext) extends Serializable with Log
     if (Metrics.isRecording) records.instrument() else records
     val samRecordConverter = new SAMRecordConverter
 
-    AlignedReadRDD(records.map(p => samRecordConverter.convert(p._2.get)),
+    AlignmentRecordRDD(records.map(p => samRecordConverter.convert(p._2.get)),
       seqDict,
       readGroups)
   }
@@ -661,7 +657,7 @@ class ADAMContext(@transient val sc: SparkContext) extends Serializable with Log
     }))
     if (Metrics.isRecording) records.instrument() else records
     val samRecordConverter = new SAMRecordConverter
-    AlignedReadRDD(records.map(p => samRecordConverter.convert(p._2.get)),
+    AlignmentRecordRDD(records.map(p => samRecordConverter.convert(p._2.get)),
       seqDict,
       readGroups)
   }
@@ -772,7 +768,7 @@ class ADAMContext(@transient val sc: SparkContext) extends Serializable with Log
     // convert avro to sequence dictionary
     val rgd = loadAvroReadGroupMetadata(filePath)
 
-    AlignedReadRDD(rdd, sd, rgd)
+    AlignmentRecordRDD(rdd, sd, rgd)
   }
 
   /**
@@ -800,7 +796,7 @@ class ADAMContext(@transient val sc: SparkContext) extends Serializable with Log
 
     // convert records
     val fastqRecordConverter = new FastqRecordConverter
-    UnalignedReadRDD.fromRdd(records.flatMap(fastqRecordConverter.convertPair))
+    AlignmentRecordRDD.unaligned(records.flatMap(fastqRecordConverter.convertPair))
   }
 
   /**
@@ -876,7 +872,7 @@ class ADAMContext(@transient val sc: SparkContext) extends Serializable with Log
       case ValidationStringency.SILENT =>
     }
 
-    UnalignedReadRDD.fromRdd(reads1.rdd ++ reads2.rdd)
+    AlignmentRecordRDD.unaligned(reads1.rdd ++ reads2.rdd)
   }
 
   /**
@@ -911,7 +907,7 @@ class ADAMContext(@transient val sc: SparkContext) extends Serializable with Log
 
     // convert records
     val fastqRecordConverter = new FastqRecordConverter
-    UnalignedReadRDD.fromRdd(records.map(
+    AlignmentRecordRDD.unaligned(records.map(
       fastqRecordConverter.convertRead(
         _,
         recordGroupOpt.map(recordGroup =>
@@ -1527,11 +1523,10 @@ class ADAMContext(@transient val sc: SparkContext) extends Serializable with Log
     } else if (filePath.endsWith(".fa") ||
       filePath.endsWith(".fasta")) {
       log.info(s"Loading $filePath as FASTA and converting to AlignmentRecords. Projection is ignored.")
-      UnalignedReadRDD(loadFasta(filePath, fragmentLength = 10000).toReads,
-        RecordGroupDictionary.empty)
+      AlignmentRecordRDD.unaligned(loadFasta(filePath, fragmentLength = 10000).toReads)
     } else if (filePath.endsWith("contig.adam")) {
       log.info(s"Loading $filePath as Parquet of NucleotideContigFragment and converting to AlignmentRecords. Projection is ignored.")
-      UnalignedReadRDD(loadParquetContigFragments(filePath).toReads, RecordGroupDictionary.empty)
+      AlignmentRecordRDD.unaligned(loadParquetContigFragments(filePath).toReads)
     } else {
       log.info(s"Loading $filePath as Parquet of AlignmentRecords.")
       loadParquetAlignments(filePath, None, projection)

--- a/adam-core/src/main/scala/org/bdgenomics/adam/rdd/fragment/FragmentRDD.scala
+++ b/adam-core/src/main/scala/org/bdgenomics/adam/rdd/fragment/FragmentRDD.scala
@@ -28,10 +28,8 @@ import org.bdgenomics.adam.models.{
 }
 import org.bdgenomics.adam.rdd.{ AvroReadGroupGenomicRDD, JavaSaveArgs }
 import org.bdgenomics.adam.rdd.read.{
-  AlignedReadRDD,
   AlignmentRecordRDD,
-  MarkDuplicates,
-  UnalignedReadRDD
+  MarkDuplicates
 }
 import org.bdgenomics.adam.serialization.AvroSerializer
 import org.bdgenomics.formats.avro._
@@ -119,11 +117,7 @@ case class FragmentRDD(rdd: RDD[Fragment],
     val newRdd = rdd.flatMap(converter.convertFragment)
 
     // are we aligned?
-    if (sequences.isEmpty) {
-      UnalignedReadRDD(newRdd, recordGroups)
-    } else {
-      AlignedReadRDD(newRdd, sequences, recordGroups)
-    }
+    AlignmentRecordRDD(newRdd, sequences, recordGroups)
   }
 
   /**

--- a/adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/AlignmentRecordRDD.scala
+++ b/adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/AlignmentRecordRDD.scala
@@ -83,7 +83,7 @@ private[adam] class AlignmentRecordArraySerializer extends IntervalArraySerializ
   }
 }
 
-sealed trait AlignmentRecordRDD extends AvroReadGroupGenomicRDD[AlignmentRecord, AlignmentRecordRDD] {
+abstract class AlignmentRecordRDD extends AvroReadGroupGenomicRDD[AlignmentRecord, AlignmentRecordRDD] {
 
   protected def buildTree(rdd: RDD[(ReferenceRegion, AlignmentRecord)])(
     implicit tTag: ClassTag[AlignmentRecord]): IntervalArray[ReferenceRegion, AlignmentRecord] = {

--- a/adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/AlignmentRecordRDD.scala
+++ b/adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/AlignmentRecordRDD.scala
@@ -83,12 +83,19 @@ private[adam] class AlignmentRecordArraySerializer extends IntervalArraySerializ
   }
 }
 
-abstract class AlignmentRecordRDD extends AvroReadGroupGenomicRDD[AlignmentRecord, AlignmentRecordRDD] {
+object AlignmentRecordRDD {
 
-  protected def buildTree(rdd: RDD[(ReferenceRegion, AlignmentRecord)])(
-    implicit tTag: ClassTag[AlignmentRecord]): IntervalArray[ReferenceRegion, AlignmentRecord] = {
-    IntervalArray(rdd, AlignmentRecordArray.apply(_, _))
+  def unaligned(rdd: RDD[AlignmentRecord]): AlignmentRecordRDD = {
+    AlignmentRecordRDD(rdd,
+      SequenceDictionary.empty,
+      RecordGroupDictionary.empty)
   }
+}
+
+case class AlignmentRecordRDD(
+    rdd: RDD[AlignmentRecord],
+    sequences: SequenceDictionary,
+    recordGroups: RecordGroupDictionary) extends AvroReadGroupGenomicRDD[AlignmentRecord, AlignmentRecordRDD] {
 
   /**
    * Replaces the underlying RDD and SequenceDictionary and emits a new object.
@@ -98,7 +105,20 @@ abstract class AlignmentRecordRDD extends AvroReadGroupGenomicRDD[AlignmentRecor
    * @return Returns a new AlignmentRecordRDD.
    */
   protected def replaceRddAndSequences(newRdd: RDD[AlignmentRecord],
-                                       newSequences: SequenceDictionary): AlignmentRecordRDD
+                                       newSequences: SequenceDictionary): AlignmentRecordRDD = {
+    AlignmentRecordRDD(newRdd,
+      newSequences,
+      recordGroups)
+  }
+
+  protected def replaceRdd(newRdd: RDD[AlignmentRecord]): AlignmentRecordRDD = {
+    copy(rdd = newRdd)
+  }
+
+  protected def buildTree(rdd: RDD[(ReferenceRegion, AlignmentRecord)])(
+    implicit tTag: ClassTag[AlignmentRecord]): IntervalArray[ReferenceRegion, AlignmentRecord] = {
+    IntervalArray(rdd, AlignmentRecordArray.apply(_, _))
+  }
 
   /**
    * Convert this set of reads into fragments.
@@ -930,56 +950,3 @@ abstract class AlignmentRecordRDD extends AvroReadGroupGenomicRDD[AlignmentRecor
     replaceRdd(finalRdd)
   }
 }
-
-case class AlignedReadRDD(rdd: RDD[AlignmentRecord],
-                          sequences: SequenceDictionary,
-                          recordGroups: RecordGroupDictionary) extends AlignmentRecordRDD {
-
-  protected def replaceRddAndSequences(newRdd: RDD[AlignmentRecord],
-                                       newSequences: SequenceDictionary): AlignmentRecordRDD = {
-    AlignedReadRDD(newRdd,
-      newSequences,
-      recordGroups)
-  }
-
-  protected def replaceRdd(newRdd: RDD[AlignmentRecord]): AlignedReadRDD = {
-    copy(rdd = newRdd)
-  }
-}
-
-object UnalignedReadRDD {
-
-  /**
-   * Creates an unaligned read RDD where no record groups are attached.
-   *
-   * @param rdd RDD of reads.
-   * @return Returns an unaligned read RDD with an empty record group dictionary.
-   */
-  private[rdd] def fromRdd(rdd: RDD[AlignmentRecord]): UnalignedReadRDD = {
-    UnalignedReadRDD(rdd, RecordGroupDictionary.empty)
-  }
-}
-
-case class UnalignedReadRDD(rdd: RDD[AlignmentRecord],
-                            recordGroups: RecordGroupDictionary) extends AlignmentRecordRDD
-    with Unaligned {
-
-  protected def replaceRdd(newRdd: RDD[AlignmentRecord]): UnalignedReadRDD = {
-    copy(rdd = newRdd)
-  }
-
-  protected def replaceRddAndSequences(newRdd: RDD[AlignmentRecord],
-                                       newSequences: SequenceDictionary): AlignmentRecordRDD = {
-
-    // are we still unaligned?
-    if (newSequences.isEmpty) {
-      UnalignedReadRDD(newRdd,
-        recordGroups)
-    } else {
-      AlignedReadRDD(newRdd,
-        newSequences,
-        recordGroups)
-    }
-  }
-}
-

--- a/adam-core/src/test/scala/org/bdgenomics/adam/rdd/read/AlignmentRecordRDDSuite.scala
+++ b/adam-core/src/test/scala/org/bdgenomics/adam/rdd/read/AlignmentRecordRDDSuite.scala
@@ -72,7 +72,7 @@ class AlignmentRecordRDDSuite extends ADAMFunSuite {
     val contigNames = rdd.flatMap(r => Option(r.getContigName)).distinct.collect
     val sd = new SequenceDictionary(contigNames.map(v => SequenceRecord(v, 1000000L)).toVector)
 
-    val sortedReads = AlignedReadRDD(rdd, sd, RecordGroupDictionary.empty)
+    val sortedReads = AlignmentRecordRDD(rdd, sd, RecordGroupDictionary.empty)
       .sortReadsByReferencePosition()
       .rdd
       .collect()
@@ -151,7 +151,7 @@ class AlignmentRecordRDDSuite extends ADAMFunSuite {
       }).toVector)
 
     val rdd = sc.parallelize(reads)
-    val sortedReads = AlignedReadRDD(rdd, sd, RecordGroupDictionary.empty)
+    val sortedReads = AlignmentRecordRDD(rdd, sd, RecordGroupDictionary.empty)
       .sortReadsByReferencePositionAndIndex()
       .rdd
       .collect()

--- a/adam-core/src/test/scala/org/bdgenomics/adam/rdd/read/MarkDuplicatesSuite.scala
+++ b/adam-core/src/test/scala/org/bdgenomics/adam/rdd/read/MarkDuplicatesSuite.scala
@@ -96,7 +96,7 @@ class MarkDuplicatesSuite extends ADAMFunSuite {
   }
 
   private def markDuplicates(reads: AlignmentRecord*): Array[AlignmentRecord] = {
-    AlignedReadRDD(sc.parallelize(reads), SequenceDictionary.empty, rgd)
+    AlignmentRecordRDD(sc.parallelize(reads), SequenceDictionary.empty, rgd)
       .markDuplicates()
       .rdd
       .collect()
@@ -205,7 +205,7 @@ class MarkDuplicatesSuite extends ADAMFunSuite {
   }
 
   private def markDuplicateFragments(reads: AlignmentRecord*): Array[AlignmentRecord] = {
-    AlignedReadRDD(sc.parallelize(reads), SequenceDictionary.empty, rgd)
+    AlignmentRecordRDD(sc.parallelize(reads), SequenceDictionary.empty, rgd)
       .toFragments
       .markDuplicates()
       .toReads

